explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,260 @@
+
+If everything went well, we can now use the `gcov` file and feed it back to GCC when recompiling OpenSSL.
+
+### Rebuilding OpenSSL with PGO
+
+We have everything we need to rebuild OpenSSL and make use of our profile data.  The most important thing now is to set the `CFLAGS` environment variable properly so that GCC can find and use the profile we generated in the previous step.
+
+How you perform this `CFLAGS` adjustment depends on how you built the software in the first place, so we won't cover this part here.  The resulting `CFLAGS` variable should have the following GCC option, though:
+
+```
+-fauto-profile=/path/to/openssl-nonpgo.gcov
+```
+
+Make sure to adjust the path to the `openssl-nonpgo.gcov` file.
+
+Also note that enabling PGO will make GCC automatically enable several optimization passes that are usually disabled, even when using `-O2`.  This can lead to new warnings/errors in the code, which might require either the code to be adjusted or the use of extra compilation flags to suppress these warnings.
+
+After the software is built, you can install this new version and run the benchmark tests again to compare results.
+
+### Our results
+
+At first sight it might appear that our results don't have much to show for themselves, but they are actually very interesting and allow us to discuss some more things about PGO.
+
+First, it is not a silver bullet.  It will not work for every scenario, and there might be cases when it is not justified to use it.  For example, if you are dealing with a very generic program that doesn't have a single representative workload.  PGO is great when applied to specialized cases, but may negatively impact the performance of programs that can be used in a variety of ways.
+
+It is also really important to be rigorous when doing the profiling of the program you want to optimize.  Here are a few tips that might help obtain more confident results:
+
+* Make sure to perform multiple runs of the same workload.  This helps to eliminate possible outliers as well as decrease the contribution of external factors (like memory and cache times).  In our example, we performed 5 consecutive runs without PGO and 5 consecutive runs with PGO.  Had we performed more runs, some of the outliers we see in the results would be even less noticeable.
+
+* Try to configure your system to have as little load as possible.  If you have other programs running in the background, they will certainly affect the final measurements in non-deterministic ways.
+
+* Another important thing to consider is whether you are dealing with code that has already been highly optimized.  In fact, this is exactly the case with our example: OpenSSL is an ubiquitous cryptographic library that has received *a lot* of attention from several developers, and it is expected that cryptographic libraries are extremely optimized.  This is the major factor that explains why our results were not so great: the library is already pretty performant as it is!
```

Comment:
```suggestion
* Consider whether you are dealing with code that has already been highly optimized.  In fact, this is exactly the case with our example: OpenSSL is a ubiquitous cryptographic library that has received *a lot* of attention from several developers, and it is expected that cryptographic libraries are extremely optimized.  This is the major factor that explains why our results were not so great: the library is already pretty performant as it is!
```

Reply:
I'm not sure it's fair to say that the results are "not so great" - I'd suggest framing it more as that you saw minimal performance gains or something like that.

Reply:
Fair point.  I've rewritten that part.